File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/wallet/CroCardWallet.java
package at.msd.friehs_bicha.cdcsvparser.wallet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;

import at.msd.friehs_bicha.cdcsvparser.App.CroCardTxApp;
import at.msd.friehs_bicha.cdcsvparser.transactions.CroCardTransaction;
import at.msd.friehs_bicha.cdcsvparser.transactions.Transaction;

public class CroCardWallet extends Wallet implements Serializable {

    public static ArrayList<String> tts = new ArrayList<>();
    String transactionType;
    CroCardTxApp txApp;

    public CroCardWallet(String currencyType, BigDecimal amount, String transactionType, CroCardTxApp txApp) {
        super(currencyType, amount, amount);
        this.transactionType = transactionType;
        if (!tts.contains(transactionType)) {
            tts.add(transactionType);
        }
        this.txApp = txApp;
    }


    public void addTransaction(Transaction transaction) {

        CroCardTransaction cardTransaction = (CroCardTransaction) transaction;

        String tt = (cardTransaction.getTransactionTypeString());
        //tt = checkForRefund(tt);

        if (tt.equals("EUR -> EUR")) {
            System.out.println("Found EUR -> EUR: " + tt);
        }

        if (tts.contains(tt)) {
            CroCardWallet w = findWallet(tt);
            if (!txApp.isUseStrictWalletType())
            {
                w = getNonStrictWallet(tt);
            }
            w.addToWallet(transaction);
            w.transactions.add(cardTransaction);
        } else {
            CroCardWallet w;
            if (!txApp.isUseStrictWalletType())
            {
                w = getNonStrictWallet(tt);
                if (w == null){
                    w = new CroCardWallet("EUR", cardTransaction.getAmount(), tt, txApp);
                    txApp.wallets.add(w);
                    w.transactions.add(cardTransaction);
                } else {
                    w.addToWallet(transaction);
                }
            } else {
                w = new CroCardWallet("EUR", cardTransaction.getAmount(), tt, txApp);
                w.transactions.add(cardTransaction);
                txApp.wallets.add(w);
            }
            //w.addToWallet(transaction.getAmount());

        }
    }


    /**
     * Get Wallet index from CurrencyType String
     *
     * @param ct the CurrencyType as String
     * @return the index of the wallet
     */
    public int getWallet(String ct) {
        int i = 0;
        for (Wallet w : txApp.wallets) {
            if (((CroCardWallet)w).transactionType.equals(ct)) return i;
            i++;
        }
        return -1;
    }


    public void writeAmount() {
        BigDecimal amountSpent = BigDecimal.ZERO;
        for (Wallet w : txApp.wallets) {
            //System.out.println("-".repeat(20));
            System.out.println(((CroCardWallet)w).transactionType);
            System.out.println(w.amount);
            System.out.println(w.moneySpent);
            System.out.println("Transactions: " + ((CroCardWallet)w).transactions.size());
            amountSpent = amountSpent.add(w.moneySpent);
        }
        //System.out.println("-".repeat(20));
        System.out.println("Amount total spent: " + amountSpent);
    }


    public void addToWallet(Transaction transaction) {
        this.amount = this.amount.add(transaction.getAmount());
        this.moneySpent = this.moneySpent.add(transaction.getAmount());
        this.transactions.add(transaction);
    }

    public String getTransactionType() {
        return transactionType;
    }


    /**
     * Remove a transaction from the CDCWallet
     *
     * @param amount the amount to remove
     * @param nativeAmount the amount in native currency to remove
     * @deprecated use {@link #removeFromWallet(Transaction)} instead
     */
    public void removeFromWallet(BigDecimal amount, BigDecimal nativeAmount) {
        this.amount = this.amount.subtract(amount);
        this.moneySpent = this.moneySpent.subtract(nativeAmount);
    }

    /**
     * Remove a transaction from the CDCWallet
     *
     * @param transaction the transaction to remove
     */
    public void removeFromWallet(Transaction transaction) {
        this.amount = this.amount.subtract(transaction.getAmount());
        this.moneySpent = this.moneySpent.subtract(transaction.getAmount());
        this.transactions.add(transaction);
    }

    private CroCardWallet getNonStrictWallet(String tt){

        tt = checkForRefund(tt);

        for (Wallet w : txApp.wallets) {
            if (tt.contains(" ")) {
                if (((CroCardWallet) w).transactionType.contains(tt.substring(0, tt.indexOf(" ")))) {
                    ((CroCardWallet) w).transactionType = tt.substring(0, tt.indexOf(" "));
                    checkTTS(tt, tt.substring(0, tt.indexOf(" ")));
                    return ((CroCardWallet) w);
                }
            }
        }

        return null;

    }

    @NonNull
    private String checkForRefund(String tt) {
        if (tt.contains("Refund: ")){
            tt = checkTTS(tt, tt.substring(8));
        }
        if (tt.contains("Refund reversal: ")){
            tt = checkTTS(tt, tt.substring(17));
        }
        return tt;
    }


    /**
     * Check if the transaction type is already in the tts list and replace it with the new one
     *
     * @param tt the transaction type to check
     * @param txType the new transaction type
     * @return the new transaction type
     */
    private String checkTTS(String tt, String txType) {
        if (tts.contains(tt)){
            tts.remove(tt);
            tts.add(txType);
        }
        return txType;
    }

    @Nullable
    private CroCardWallet findWallet(String tt) {

        if (!txApp.isUseStrictWalletType()) {
            tt = checkForRefund(tt);
        }

        for (Wallet w : txApp.wallets) {
            if (((CroCardWallet) w).transactionType.equals(tt)) {
                return ((CroCardWallet) w);
            }
        }
        return null;
    }


}

File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/wallet/CroCardWallet.kt
package at.msd.friehs_bicha.cdcsvparser.wallet

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.TypeConverters
import at.msd.friehs_bicha.cdcsvparser.app.BaseApp
import at.msd.friehs_bicha.cdcsvparser.logging.FileLog
import at.msd.friehs_bicha.cdcsvparser.transactions.CroCardTransaction
import at.msd.friehs_bicha.cdcsvparser.transactions.CurveCardTx
import at.msd.friehs_bicha.cdcsvparser.transactions.Transaction
import at.msd.friehs_bicha.cdcsvparser.util.Converter
import java.io.Serializable
import java.math.BigDecimal

/**
 * Represents a CroCardWallet object
 */
@Entity(tableName = "card_wallets")
@TypeConverters(Converter::class, Converter.BigDecimalConverter::class)
class CroCardWallet(
    currencyType: String,
    amount: BigDecimal?,
    var transactionType: String?,
    txApp: BaseApp?
) : Wallet(currencyType, amount, amount), Serializable {
    constructor(
        walletID: Long,
        currencyType: String,
        amount: Double?,
        amountBonus: Double?,
        moneySpent: Double,
        outsideWallet: Boolean,
        transactions: MutableList<CroCardTransaction?>,
        transactionType: String?
    ) : this(
        currencyType,
        amount?.let { BigDecimal(it) },
        currencyType,
        null
    ) {
        this.walletId = walletID.toInt()
        this.currencyType = currencyType
        this.transactionType = transactionType
        amount?.let { this.amount = BigDecimal(it) }
        amountBonus?.let { this.amountBonus = BigDecimal(it) }
        this.moneySpent = BigDecimal(moneySpent)
        this.transactions = transactions as MutableList<Transaction?>
        this.isOutsideWallet = outsideWallet
    }

    //constructor for all members:
    constructor(
        walletId: Int,
        currencyType: String,
        amount: BigDecimal?,
        amountBonus: BigDecimal?,
        moneySpent: BigDecimal,
        isOutsideWallet: Boolean,
        transactionType: String?
    ) : this(
        currencyType,
        amount,
        transactionType,
        null
    ) {
        this.walletId = walletId
        this.currencyType = currencyType
        this.transactionType = transactionType
        this.amount = amount!!
        this.amountBonus = amountBonus!!
        this.moneySpent = moneySpent
        //if(transactions != null) this.transactions = transactions as MutableList<Transaction?>
        this.isOutsideWallet = isOutsideWallet
    }

    @Ignore
    lateinit var txApp: BaseApp

    init {
        if (!tts.contains(transactionType)) {
            tts.add(transactionType)
        }
        if (txApp != null) {
            this.txApp = txApp
        }
    }

    override fun addTransaction(transaction: Transaction) {
        val cardTransaction = transaction as CroCardTransaction
        val tt = cardTransaction.transactionTypeString
        if (tt == "EUR -> EUR") {
            FileLog.i("CCW.addTx","Found EUR -> EUR: $tt")
        }

        var ignoreThisTx = false

        if (transaction is CurveCardTx) {
            if (transaction.txType == "REFUNDED") {
                FileLog.i("CCW.addTx", "Ignoring CurveCardTx: $transaction")
                ignoreThisTx = true
            }
            if (transaction.txType == "Declined") {
                FileLog.i("CCW.addTx", "Ignoring CurveCardTx: $transaction")
                ignoreThisTx = true
            }
        }

        var w = findWallet(tt)
        if (w != null) {
            w = findWallet(tt)
            if (!txApp.isUseStrictWalletType) {
                w = getNonStrictWallet(tt)
            }
            if (w == null) {
                w = CroCardWallet("EUR", BigDecimal.ZERO, tt, txApp)
            }
            w.addToWallet(transaction, ignoreThisTx)
            w.transactions.add(cardTransaction)
            transaction.walletId = w.walletId
        } else {
            if (!txApp.isUseStrictWalletType) {
                w = getNonStrictWallet(tt)
                if (w == null) {
                    if (!ignoreThisTx) w = CroCardWallet("EUR", cardTransaction.amount, tt, txApp)
                    else w = CroCardWallet("EUR", BigDecimal.ZERO, tt, txApp)
                    txApp.wallets.add(w)
                    w.transactions.add(cardTransaction)
                    transaction.walletId = w.walletId
                } else {
                    w.addToWallet(transaction, ignoreThisTx)
                    transaction.walletId = w.walletId
                }
            } else {
                w = CroCardWallet("EUR", cardTransaction.amount, tt, txApp)
                w.transactions.add(cardTransaction)
                transaction.walletId = w.walletId
                txApp.wallets.add(w)
            }
        }
    }

    /**
     * Get Wallet index from CurrencyType String
     *
     * @param ct the CurrencyType as String
     * @return the index of the wallet
     */
    override fun getWallet(ct: String?): Int {
        for ((i, w) in txApp.wallets.withIndex()) {
            if ((w as CroCardWallet).transactionType == ct) return i
        }
        return -1
    }

    fun addToWallet(transaction: Transaction, ignoreThisTx: Boolean = false) {
        transaction.walletId = walletId
        if (ignoreThisTx) {
            FileLog.i("CCW.addToWallet", "Ignoring transaction: $transaction")
            transactions.add(transaction)
            return
        }
        if (transaction.currencyType != "EUR") {
            amount = amount.add(transaction.nativeAmount)
            moneySpent = moneySpent.add(transaction.nativeAmount)
            transactions.add(transaction)
        } else {
            amount = amount.add(transaction.amount)
            moneySpent = moneySpent.add(transaction.amount)
            transactions.add(transaction)
        }
    }

    /**
     * Remove a transaction from the CDCWallet
     *
     * @param amount       the amount to remove
     * @param nativeAmount the amount in native currency to remove
     */
    @Deprecated("use {@link #removeFromWallet(Transaction)} instead")
    override fun removeFromWallet(amount: BigDecimal?, nativeAmount: BigDecimal?) {
        this.amount = this.amount.subtract(amount)
        moneySpent = moneySpent.subtract(nativeAmount)
    }

    /**
     * Remove a transaction from the CDCWallet
     *
     * @param transaction the transaction to remove
     */
    fun removeFromWallet(transaction: Transaction) {
        amount = amount.subtract(transaction.amount)
        moneySpent = moneySpent.subtract(transaction.amount)
        transactions.add(transaction)
    }

    private fun getNonStrictWallet(tt: String?): CroCardWallet? {
        var tt = tt!!
        tt = checkForRefund(tt)
        for (w in txApp.wallets) {
            if (tt.contains(" ")) {
                if ((w as CroCardWallet).transactionType!!.contains(
                        tt.substring(
                            0,
                            tt.indexOf(" ")
                        )
                    )
                ) {
                    w.transactionType = tt.substring(0, tt.indexOf(" "))
                    checkTTS(tt, tt.substring(0, tt.indexOf(" ")))
                    return w
                }
            } else if ((w as CroCardWallet).transactionType!!.contains(tt)) {
                return w
            }
        }
        return null
    }

    private fun checkForRefund(tt: String): String {
        var tt = tt
        if (tt.contains("Refund: ")) {
            tt = checkTTS(tt, tt.substring(8))
        }
        if (tt.contains("Refund reversal: ")) {
            tt = checkTTS(tt, tt.substring(17))
        }
        return tt
    }

    /**
     * Check if the transaction type is already in the tts list and replace it with the new one
     *
     * @param tt     the transaction type to check
     * @param txType the new transaction type
     * @return the new transaction type
     */
    private fun checkTTS(tt: String, txType: String): String {
        if (tts.contains(tt)) {
            tts.remove(tt)
        }
        tts.add(txType)
        return txType
    }

    private fun findWallet(tt: String?): CroCardWallet? {
        var tt = tt!!
        if (!txApp.isUseStrictWalletType) {
            tt = checkForRefund(tt)
        }
        for (w in txApp.wallets) {
            if ((w as CroCardWallet).transactionType == tt) {
                return w
            }
        }
        return null
    }

    override fun getTypeString(): String {
        return transactionType!!
    }

    companion object {
        var tts = ArrayList<String?>()

        /**
         * Converts a HashMap<String, *> to a CroCardWallet object
         *
         * @param wallet
         * @return
         */
        fun fromDb(wallet: HashMap<String, *>): CroCardWallet {
            val walletId = wallet["walletId"] as Long
            val currencyType = wallet["currencyType"] as String
            val amount = wallet["amount"] as Double
            val amountBonus = wallet["amountBonus"] as Double
            val moneySpent = wallet["moneySpent"] as Double
            val isOutsideWallet = wallet["outsideWallet"] as Boolean
            val transactionType = wallet["transactionType"] as String

            val transactionsList = wallet["transactions"] as MutableList<java.util.HashMap<String, *>?>?
            val transactions = ArrayList<CroCardTransaction?>()

            transactionsList?.forEach { transactionMap ->
                transactionMap?.let {
                    transactions.add(CroCardTransaction.fromDb(it))
                }
            }

            return CroCardWallet(
                walletId,
                currencyType,
                amount,
                amountBonus,
                moneySpent,
                isOutsideWallet,
                transactions,
                transactionType
            )
        }
    }
}